libavfilter/sf_text2graphicsub.c

```diff
@@ -0,0 +1,526 @@
+        style->BackColour       = 0x00000080;
+        style->Bold             = 200;
+        style->ScaleX           = 1.0;
+        style->ScaleY           = 1.0;
+        style->Spacing          = 0;
+        style->BorderStyle      = 1;
+        style->Outline          = 2;
+        style->Shadow           = 3;
+        style->Alignment        = 2;
+        track->default_style = sid;
+    }
+
+    s->got_header = 1;
+}
+
+// Main interface functions
+
+static av_cold int init(AVFilterContext *ctx)
+{
+    Text2GraphicSubContext *context = ctx->priv;
+
+    context->library = ass_library_init();
+    ass_set_message_cb(context->library, ass_log, NULL);
+    ass_set_fonts_dir(context->library, context->fontsdir);
+    ass_set_extract_fonts(context->library, 1);
+
+    if (context->filename) {
+        AVFormatContext *video = NULL;
+        int ret = avformat_open_input(&video, context->filename, NULL, NULL);
+        if (ret < 0) {
+            av_log(ctx, AV_LOG_ERROR, "Unable to open %s\n", context->filename);
+            return ret;
+        }
+        for (int i = 0; i < video->nb_streams; i++) {
+            const AVDictionaryEntry *tag;
+            AVStream *st = video->streams[i];
+            if (!stream_is_font(st)) continue;
+            tag = av_dict_get(st->metadata, "filename", NULL, AV_DICT_MATCH_CASE);
+            if (!tag) continue;
+            av_log(NULL, AV_LOG_DEBUG, "Loading attached font: %s\n", tag->value);
+            ass_add_font(context->library, tag->value,
+                 (char *)st->codecpar->extradata,
+                 st->codecpar->extradata_size);
+        }
+        avformat_close_input(&video);
+    }
+
+    context->renderer = ass_renderer_init(context->library);
+    ass_set_pixel_aspect(context->renderer, 1);
+    ass_set_shaper(context->renderer, 0);
+    ass_set_fonts(context->renderer, NULL, NULL, 1, NULL, 1);
+
+    context->track = ass_new_track(context->library);
+    if (!context->track) {
+        av_log(ctx, AV_LOG_ERROR, "ass_new_track() failed!\n");
+        return AVERROR(EINVAL);
+    }
+
+    ass_set_check_readorder(context->track, 0);
+
+    init_palettizecontext(&context->palettize_context);
+
+    ff_draw_init(&context->draw_context, AV_PIX_FMT_BGRA, FF_DRAW_PROCESS_ALPHA);
+
+    return 0;
+}
+
+static int config_input(AVFilterLink *inlink)
+{
+    const AVFilterContext *ctx = inlink->dst;
+    Text2GraphicSubContext *context = ctx->priv;
+    if (!context->size.width)  context->size.width  = inlink->w;
+    if (!context->size.height) context->size.height = inlink->h;
+    if (!context->size.height || !context->size.width) {
+        av_log(NULL, AV_LOG_ERROR, "A positive height and width are required to render subtitles\n");
+        return AVERROR_EXIT;
+    }
+    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
+    ass_set_storage_size(context->renderer, context->size.width, context->size.height);
```

Comment:
We need to change this line to

```
    ass_set_storage_size(context->renderer, inlink->w, inlink->h);
```

This must be the input size, which is the size for which the ASS subs were originally created for.
libass uses this to properly rescale values like font size, outline size, shadow depths and other things to preserve the appearance even the target size is different from the original size.